bftengine/src/bftengine/SimpleClientImp.cpp

```diff
@@ -204,173 +225,291 @@ SimpleClientImp::SimpleClientImp(ICommunication* communication,
   communication_->setReceiver(clientId_, this);
 }
 
-SimpleClientImp::~SimpleClientImp() {
-  ConcordAssert(replysCertificate_.isEmpty());
+SimpleClientImp::~SimpleClientImp() {}
+
+bool SimpleClientImp::allRequiredRepliesReceived() {
+  if (replysCertificate_.empty()) return false;
+  for (auto& elem : replysCertificate_) {
+    if (!elem.second->isComplete()) {
+      LOG_DEBUG(logger_, "Not all replies received" << KVLOG(clientId_));
+      return false;
+    }
+  }
+  LOG_DEBUG(logger_, "All replies received" << KVLOG(clientId_));
+  return true;
+}
+
+void SimpleClientImp::verifySendRequestPrerequisites() {
+  ConcordAssert(replysCertificate_.empty());
   ConcordAssert(msgQueue_.empty());
-  ConcordAssert(pendingRequest_ == nullptr);
+  ConcordAssert(pendingRequest_.empty());
   ConcordAssert(timeOfLastTransmission_ == MinTime);
   ConcordAssert(numberOfTransmissions_ == 0);
 }
 
+tuple<bool, bool> SimpleClientImp::pendingRequestProcessing(bool isBatch,
+                                                            const Time& beginTime,
+                                                            uint64_t reqTimeoutMilli,
+                                                            const std::string& cid) {
+  bool requestCommitted = false;
+  bool requestTimedOut = false;
+  auto predicate = [this] { return !msgQueue_.empty(); };
```

Comment:
`msgQueueNonEmpty` instead` of `predicate`?

Reply:
np